src/libstd/io/buffered.rs

```diff
@@ -451,7 +453,14 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => unsafe {
+                // use `ptr::read` to extract the inner writer and circumvent
+                // our `Drop` implementation
+                mem::replace(&mut self.buf, Vec::new());
```

Comment:
Maybe `std::ptr::drop_in_place(&mut self.buf)` would be the clearest then, but I still want to favour safe code when it should be equivalent.